tensorflow/c/kernels.cc

```diff
@@ -66,6 +66,42 @@ void TF_DeleteKernelBuilder(TF_KernelBuilder* builder) {
   }
 }
 
+typedef std::function<void()> DoneCallback;
+void TF_RunDoneCallback(TF_DoneCallback* done) {
+  (*reinterpret_cast<DoneCallback*>(done))();
+}
+
+struct TF_AsyncKernelBuilder {
```

Comment:
@sagunb @chuanhaozhuge Do you think we should just reuse struct `TF_KernelBuilder` instead? Its members aren't exposed anyway. (And add another member for `compute_async_function`)?

This way, we don't need `TF_DeleteAsyncKernelBuilder`, `TF_AsyncKernelBuilder_TypeConstraint`, `TF_AsyncKernelBuilder_HostMemory`, and `TF_AsyncKernelBuilder_Priority`. (Can just reuse the old API because they have nothing to do with asynchronicity.)

Reply:
The drawback of reusing struct `TF_KernelBuilder` is,  `TF_NewKernelBuilder` needs another argument `compute_async_function`, that will break the compatibility of this API. I dont have strong opinions on this.

Reply:
This is what I'd like to check with @sagunb about. `TF_KernelBuilder` is opaque to external code. (Its definition is in kernels.cc, not kernels.h). External codes are supposed to store the `TF_KernelBuilder` pointer returned from `TF_NewKernelBuilder` and pass it to other C API to extract values. External codes will break if they somehow try to extract values directly from the pointer, or have some dependency on struct size. But I don't think TF guarantees anything about opaque structs in the first place.

@yisitu May I hear your thoughts on this too? :)

Reply:
@sagunb @chuanhaozhuge @yisitu @penpornk Any updates on this issue?

Reply:
Sorry for the super delayed replay. I prefer using a new `TF_AsyncKernelBuilder` so that `TF_NewKernelBuilder` can stay intact from this change.

Reply:
I think it's better to re-use TF_KernelBuilder with a compute_async_function member. @penpornk is right, TF_KernelBuilder itself is opaque and not part of the public API.

We would still need to add a new `TF_NewAsyncKernelBuilder` function in kernel.h to provide a way to specify that this KernelBuilder is async but we would not need all the other associated functions such as TF_DeleteAsyncKernelBuilder, TF_AsyncKernelBuilder_TypeConstraint, TF_AsyncKernelBuilder_HostMemory, and TF_AsyncKernelBuilder_Priority

Reply:
@sagunb Thanks for your comments. Done.